pyccel/parser/syntax/openmp.py

```diff
@@ -11,396 +11,962 @@
+class OmpTaskWaitConstruct(OmpConstruct):
+    """Class representing a TaskWait construct."""
+    def __init__(self, **kwargs):
+        super().__init__(OmpAnnotatedComment, None, **kwargs)
 
-    @property
-    def expr(self):
-        if DEBUG:
-            print("> OmpSingleConstruct: expr")
+class OmpTaskyieldConstruct(OmpConstruct):
+    """Class representing a Taskyield construct."""
+    def __init__(self, **kwargs):
+        super().__init__(OmpAnnotatedComment, None, **kwargs)
 
-        _valid_clauses = (OmpPrivate, \
-                         OmpFirstPrivate)
+class OmpFlushConstruct(OmpConstruct):
+    """Class representing a Flush construct."""
+    def __init__(self, **kwargs):
+        super().__init__(OmpAnnotatedComment, (FlushList,), **kwargs)
 
-        txt = 'single'
-        for clause in self.clauses:
-            if isinstance(clause, _valid_clauses):
-                txt = '{0} {1}'.format(txt, clause.expr)
-            else:
-                raise TypeError('Wrong clause for OmpSingleConstruct')
+class OmpCancelConstruct(OmpConstruct):
+    """Class representing a Cancel construct."""
+    def __init__(self, **kwargs):
+        super().__init__(OMP_Cancel_Construct, (OmpCancelType,), **kwargs)
+
+class OmpTargetConstruct(OmpConstruct):
+    """Class representing a Target construct."""
+    def __init__(self, **kwargs):
+        super().__init__(OMP_Target_Construct, _valid_target_clauses, **kwargs)
+
+class OmpTeamsConstruct(OmpConstruct):
+    """Class representing a Teams construct."""
+    def __init__(self, **kwargs):
+        super().__init__(OMP_Teams_Construct, _valid_teams_clauses, **kwargs)
 
-        return OMP_Single_Construct(txt)
+class OmpAtomicConstruct(OmpConstruct):
+    """Class representing an Atomic construct ."""
+    def __init__(self, **kwargs):
+        super().__init__(OmpAnnotatedComment, _valid_atomic_clauses, **kwargs)
 
 class OmpEndClause(BasicStmt):
-    """Class representing a ."""
+    """Class representing an end construct."""
     def __init__(self, **kwargs):
-        """
-        """
         self.construct = kwargs.pop('construct')
         self.simd      = kwargs.pop('simd', '')
         self.nowait    = kwargs.pop('nowait', '')
 
-        super(OmpEndClause, self).__init__(**kwargs)
+        construct = ' '.join(self.construct)
+        txt = 'end {0} {1} {2}'.format(construct, self.simd, self.nowait)
+
+        self._expr = Omp_End_Clause(txt)
+
+        super().__init__(**kwargs)
+
+    @property
+    def expr(self):
+        if DEBUG:
+            print("> {}: expr".format(type(self).__name__))
+
+        return self._expr
+
+class OmpFinal(BasicStmt):
+    """Class representing a final clause"""
+    def __init__(self, **kwargs):
+        self.final = kwargs.pop('final')
+
+        super().__init__(**kwargs)
 
     @property
     def expr(self):
         if DEBUG:
-            print("> OmpEndClause: expr")
+            print("> {}: expr".format(type(self).__name__))
```

Comment:
If you do not expect anything to inherit from this class then `print("> OmpEndClause: expr")` is sufficient. I suggested `print("> {}: expr".format(type(self).__name__))` so that you could use inheritance to avoid writing it for every class. If `expr` needs to be defined anyway then the old code will run faster